causy/graph_model.py

```diff
@@ -0,0 +1,241 @@
+        graph=None,
+        pipeline_steps: Optional[List[PipelineStepInterface]] = None,
+    ):
+        self.graph = graph
+        self.pipeline_steps = pipeline_steps or []
+        self.pool = mp.Pool(mp.cpu_count() * 2)
+
+    def create_graph_from_data(self, data: List[Dict[str, float]]):
+        """
+        Create a graph from data
+        :param data: is a list of dictionaries
+        :return:
+        """
+        # initialize nodes
+        keys = data[0].keys()
+        nodes: Dict[str, List[float]] = {}
+
+        for key in keys:
+            nodes[key] = []
+
+        # load nodes into node dict
+        for row in data:
+            for key in keys:
+                nodes[key].append(row[key])
+
+        graph = Graph()
+        for key in keys:
+            graph.add_node(key, nodes[key])
+
+        self.graph = graph
+        return graph
+
+    def create_all_possible_edges(self):
+        """
+        Create all possible edges on a graph
+        TODO: replace me with the skeleton builders
+        :return:
+        """
+        for u in self.graph.nodes.values():
+            for v in self.graph.nodes.values():
+                if u == v:
+                    continue
+                self.graph.add_edge(u, v, {})
+
+    def execute_pipeline_steps(self):
+        """
+        Execute all pipeline_steps
+        :return: the steps taken during the step execution
+        """
+        action_history = []
+
+        for filter in self.pipeline_steps:
+            logger.info(f"Executing pipeline step {filter.__class__.__name__}")
+            if isinstance(filter, LogicStepInterface):
+                filter.execute(self.graph, self)
+                continue
+
+            result = self.execute_pipeline_step(filter)
+            action_history.append(
+                {"step": filter.__class__.__name__, "actions": result}
+            )
+
+        return action_history
+
+    def _format_yield(self, test_fn, graph, generator):
+        """
+        Format the yield for the parallel processing
+        :param test_fn: the pipeline_step test function
+        :param graph: the graph
+        :param generator: the generator object which generates the combinations
+        :return: yields the test function with its inputs
+        """
+        for i in generator:
+            yield [test_fn, [*i], graph]
+
+    def _take_action(self, results):
+        """
+        Take the actions returned by the test
+
+        In causy changes on the graph are not executed directly. Instead, the test returns an action which should be executed on the graph.
+        This is done to make it possible to execute the tests in parallel as well as to decide proactively at which point in the decisions taken by the pipeline step should be executed.
+        Actions are returned by the test and are executed on the graph. The actions are stored in the action history to make it possible to revert the actions or use them in a later step.
```

Comment:
test -> pipeline step? independence test or orientation rule? (it's not only tests)